tools/crmadmin.c

```diff
@@ -168,6 +167,185 @@ command_cb(const gchar *option_name, const gchar *optarg, gpointer data, GError
+    out->list_item(out, NULL, "Designated Controller is: %s", crm_str(dc));
+    out->end_list(out);
+
+    return pcmk_rc_ok;
+}
+
+PCMK__OUTPUT_ARGS("dc", "char *")
+static int
+dc_xml(pcmk__output_t *out, va_list args)
+{
+    char *dc = va_arg(args, char *);
+
+    xmlNodePtr node = pcmk__output_create_xml_node(out, "dc");
+    xmlSetProp(node, (pcmkXmlStr) "node_name", (pcmkXmlStr) crm_str(dc));
+
+    return pcmk_rc_ok;
+}
+
+
+PCMK__OUTPUT_ARGS("node-list", "xmlNode *")
+static int
+node_list(pcmk__output_t *out, va_list args)
+{
+    xmlNode *xml_node = va_arg(args, xmlNode *);
+    int found = 0;
+    xmlNode *node = NULL;
+    xmlNode *nodes = get_object_root(XML_CIB_TAG_NODES, xml_node);
+
+    out->begin_list(out, NULL, NULL, "Nodes");
+
+    for (node = first_named_child(nodes, XML_CIB_TAG_NODE); node != NULL;
+         node = crm_next_same_xml(node)) {
+        const char *node_type = BASH_EXPORT ? NULL :
+                     crm_element_value(node, XML_ATTR_TYPE);
+        out->message(out, "node", node_type,
+                     crm_str(crm_element_value(node, XML_ATTR_UNAME)),
+                     crm_str(crm_element_value(node, XML_ATTR_ID)));
+
+        found++;
+    }
+    // @TODO List Pacemaker Remote nodes that don't have a <node> entry
+
+    out->end_list(out);
+
+    if (found == 0) {
+        out->info(out, "No nodes configured");
+    }
+
+    return pcmk_rc_ok;
+}
+
+PCMK__OUTPUT_ARGS("node", "char *", "char *", "char *")
+static int
+node_text(pcmk__output_t *out, va_list args)
+{
+        char *type = va_arg(args, char *);
+        char *name = va_arg(args, char *);
+        char *id = va_arg(args, char *);
+
+        if (BASH_EXPORT) {
+            out->list_item(out, NULL, "export %s=%s", crm_str(name), crm_str(id));
+        } else {
+            out->list_item(out, NULL, "%s node: %s (%s)",
+                           type ? type : "member", crm_str(name), crm_str(id));
+        }
+
+        return pcmk_rc_ok;
+}
+
+PCMK__OUTPUT_ARGS("node", "char *", "char *", "char *")
+static int
+node_xml(pcmk__output_t *out, va_list args)
+{
+    char *type = va_arg(args, char *);
+    char *name = va_arg(args, char *);
+    char *id = va_arg(args, char *);
+
+    xmlNodePtr node = pcmk__output_create_xml_node(out, "node");
+    xmlSetProp(node, (pcmkXmlStr) "type", (pcmkXmlStr) (type ? type : "member"));
+    xmlSetProp(node, (pcmkXmlStr) "name", (pcmkXmlStr) crm_str(name));
+    xmlSetProp(node, (pcmkXmlStr) "id", (pcmkXmlStr) crm_str(id));
+
+    return pcmk_rc_ok;
+}
+
+static pcmk__message_entry_t fmt_functions[] = {
+    {"health", "default", health_text },
+    {"health", "xml", health_xml },
+    {"pacemakerd-health", "default", pacemakerd_health_text },
+    {"pacemakerd-health", "xml", pacemakerd_health_xml },
+    {"dc", "default", dc_text },
+    {"dc", "xml", dc_xml },
+    {"node-list", "default", node_list },
+    {"node", "default", node_text },
+    {"node", "xml", node_xml },
```

Comment:
We're going to have to do something.  This is why I previously asked if all the formatted output stuff was private API.  Some sort of prefix would be nice.  I have also tried to use "-list" as a meaningful suffix.  If it's in the message name, the message makes a list.  Otherwise, it's up to the caller to make a list if it wants one.  I have probably not been consistent with this.

Now that formatted output is spreading, I should probably audit everything for consistency.